docs/StarCellBio-Architecture.rst

```diff
@@ -0,0 +1,2332 @@
+.. contents::
+   :depth: 3.0
+..
+
+StarCellBio
+
+1. Architecture
+
+2. Code organization
+
+3. Django Server
+
+4. StarCellBio Student’s view
+
+5. StarCellBio Course Instructor’s view
+
+6. StarCellBio Labnotebook
+
+Architecture
+============
+
+StarCellBio is split in following components:
+
+-  Authentication, authorization and data storage on server
+
+-  Student’s is SPA (Single Page Application) that resides in the
+   browser and communicates with the server only to serialize or
+   de-serialize user session.
+
+-  Assignment Builder is SPA that resides in the browser where all
+   interaction happen, including running Studen’s view in preview mode.
+
+Authentication, authorization and data storage are implemented as Django
+application with minimal API exposed to the SPA’s.
+
+Student SPA loads including all the code it is necessary to perform all
+experiments. Code is that driven using JSON data structure. JSON is fed
+to the business logic encapsulated within /model folder. Business logic
+exposes underlying data while allowing for intercepting (analytics) or
+overriding behaviors (allows for relational querying without leaking
+implementation details to user interface). User interface is implemented
+in /ui folder as mix of .soy (google closure templates) and .js
+(javascript) objects.
+
+Student SPA uses simple MVC, event driven architecture. All events are
+registered upon initialization. Event handlers interact with model and
+request page to be re-rendered by changing URL hash or by invoking
+.repaint() method.
+
+User interface is build in such fashion that each HTML element carries
+sufficient information to determine which handler to call and which part
+of the model it impacts. Event handler is determined by class associated
+with the HTML element. “scb\_f\_” classes represent handlers to be
+invoked, while “scb\_s\_” classes represent element styling. This allows
+for very clean and DRY code that allows event handlers to be reused at
+different locations in the application as needed.
+
+StarCellBio uses RESTful URL’s to represents its state in the browser,
+and browser URL fully defines which view will be rendered. URL contains
+parameter view. As a matter of coding standard this name is the same as
+name of the file that renders this view in /ui folder. And name of
+template is the same name as view with .soy extension. All the view’s
+methods are in the scb namespace with the view name, and scb name space
+contains class that implements at least one methods ‘show’ that is
+invoked to render this view. Namespace also contains static method
+register with registers all required handlers for the view (scb\_f\_
+classes). This structure allows for easy code navigation and any new
+code should follow these convetions.
+
+Data structure on the client (browser)
+======================================
+
+Top level object is assignment list represented by AssignmentList, this
```

Comment:
Also, might be worthing rewording as "Top level object is a list of assignments...."